pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedAggregationOperator.java

```diff
@@ -80,46 +84,73 @@ protected IntermediateResultsBlock getNextBlock() {
           aggregationResults
               .add(new MinMaxRangePair(dictionary.getDoubleValue(0), dictionary.getDoubleValue(dictionarySize - 1)));
           break;
+        case DISTINCT:
         case DISTINCTCOUNT:
           switch (dictionary.getValueType()) {
             case INT:
               IntOpenHashSet intSet = new IntOpenHashSet(dictionarySize);
               for (int dictId = 0; dictId < dictionarySize; dictId++) {
+
+                // Enforce limits only for DISTINCT
+                if (aggregationFunction.getType() == DISTINCT && dictId == _limit) {
```

Comment:
Would be (arguably) faster if we avoid the `if` inside of this for loop? That means we will have to have separate the `DISTINCT` and `DISTINCTCOUNT` cases.